File: Native/Test/Node/IO.js
FABRIC = require('Fabric').createClient();

try
{
  FABRIC.IO.putTextFileContent("dummy", "badHandle");
}
catch(e)
{
  console.log('Bad handle error: ' + e);
}

var readOnlyFileAndFolderHandles = FABRIC.IO.queryUserFileAndFolderHandle(FABRIC.IO.forOpen, "dummy title", "txt", "testfile");

try {
  FABRIC.IO.putTextFileContent(readOnlyFileAndFolderHandles.file, "dummy");
}
catch (e) {
  console.log('No write access error: ' + e);
}

var fileAndFolderHandles = FABRIC.IO.queryUserFileAndFolderHandle(FABRIC.IO.forOpenWithWriteAccess, "dummy title", "txt", "testfile");
console.log('User file name: ' + FABRIC.IO.getFileHandleInfo(fileAndFolderHandles.file).fileName);

var relativeFileHandle = FABRIC.IO.buildFileHandleFromRelativePath(fileAndFolderHandles.folder + '/SubDir/testfile2.txt');

try {
  var dummy = FABRIC.IO.getTextFileContent(relativeFileHandle);
}
catch (e) {
  console.log('File not found error');//'e' has random handle name; don't display it
  }

FABRIC.IO.putTextFileContent(relativeFileHandle,"FileBody");
var body = FABRIC.IO.getTextFileContent(relativeFileHandle);
console.log('File content: ' + body);

FABRIC.IO.putTextFileContent(relativeFileHandle," with appended text",true);
body = FABRIC.IO.getTextFileContent(relativeFileHandle);
console.log('File content with appended text: ' + body);

var fileInfo = FABRIC.IO.getFileHandleInfo(relativeFileHandle);
console.log('File info: ' + JSON.stringify(fileInfo));

node = FABRIC.DependencyGraph.createResourceLoadNode("node");

var callbackStep = 0;

node.addOnLoadFailureCallback( function() {
  if( callbackStep == 0 ){
    console.log( 'bad url load failure' );
    node.setData('url',0,relativeFileHandle);
    callbackStep = 1;
    node.evaluate();
  }
  else
  {
    console.log('Unexpected');
    FABRIC.close();
  }
});

node.addOnLoadSuccessCallback( function() {
  console.log( 'url file load succeeded' );
  if( callbackStep == 1 ){
    var resourceResultFileHandle = FABRIC.IO.buildFileHandleFromRelativePath(fileAndFolderHandles.folder + '/SubDir/testfile3.txt');
    node.putResourceToFile(resourceResultFileHandle,'resource');
    var body = FABRIC.IO.getTextFileContent(resourceResultFileHandle);
    console.log('Resource data saved to file: ' + body);

    var sourceFileHandle = FABRIC.IO.buildFileHandleFromRelativePath(fileAndFolderHandles.folder + '/SubDir/testfile4.txt');
    FABRIC.IO.putTextFileContent(sourceFileHandle,"Resource source data from direct file");

    //Load from explicit file location, to a file handle
    callbackStep = 2;
    node.setData('storeDataAsFile',0,true);
    node.setData('url',0,'testfile://TMP/SubDir/testfile4.txt');
    node.evaluate();
  }
  else if( callbackStep == 2 ){
    var resourceResult = node.getData('resource',0);
    var resourceFileContent = FABRIC.IO.getTextFileContent(resourceResult.dataExternalLocation);
    console.log('Resource file content: ' + resourceFileContent);

    FABRIC.close();
  }
  else
  {
    console.log('Unexpected');
    FABRIC.close();
  }
});

node.setData('url',0,'badURL');
node.evaluate();

File: Native/Test/Python/IO.py
#
#  Copyright 2010-2012 Fabric Technologies Inc. All rights reserved.
#

import fabric
fabricClient = fabric.createClient()

try:
  fabricClient.IO.putTextFileContent("dummy", "badHandle")
except Exception as e:
  print('Bad handle error: ' + str(e))

readOnlyFileAndFolderHandles = fabricClient.IO.queryUserFileAndFolderHandle(fabricClient.IO.forOpen, "dummy title", "txt", "testfile")

try: 
  fabricClient.IO.putTextFileContent(readOnlyFileAndFolderHandles['file'], "dummy")
except Exception as e:
  print('No write access error: ' + str(e))


fileAndFolderHandles = fabricClient.IO.queryUserFileAndFolderHandle(fabricClient.IO.forOpenWithWriteAccess, "dummy title", "txt", "testfile")
print('User file name: ' + fabricClient.IO.getFileHandleInfo(fileAndFolderHandles['file'])['fileName'])

relativeFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile2.txt')

try:
  dummy = fabricClient.IO.getTextFileContent(relativeFileHandle)
except Exception as e:
  print('File not found error')#'e' has random handle name; don't display it

fabricClient.IO.putTextFileContent(relativeFileHandle,"FileBody")
body = fabricClient.IO.getTextFileContent(relativeFileHandle)
print('File content: ' + body)

fabricClient.IO.putTextFileContent(relativeFileHandle," with appended text",True)
body = fabricClient.IO.getTextFileContent(relativeFileHandle)
print('File content with appended text: ' + body)

fileInfo = fabricClient.IO.getFileHandleInfo(relativeFileHandle)
print('File info: ' + fabric.stringify(fileInfo))

node = fabricClient.DependencyGraph.createResourceLoadNode("node")

myvars = {}
myvars['callbackStep'] = 0

def loadFailureCallback( selfNode ):
  if myvars['callbackStep'] == 0:
    print( 'bad url load failure' )
    node.setData('url',0,relativeFileHandle)
    myvars['callbackStep'] = 1
    node.evaluate()
  else:
    print('failure callback: Unexpected')
    fabricClient.close()

node.addOnLoadFailureCallback( loadFailureCallback )

def loadSuccessCallback( selfNode ):
  print( 'url file load succeeded' )
  if myvars['callbackStep'] == 1:
    resourceResultFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile3.txt')
    node.putResourceToFile(resourceResultFileHandle,'resource')
    body = fabricClient.IO.getTextFileContent(resourceResultFileHandle)
    print('Resource data saved to file: ' + body)

    sourceFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile4.txt')
    fabricClient.IO.putTextFileContent(sourceFileHandle,"Resource source data from direct file")
    fabricClient.flush()

    #Load from explicit file location, to a file handle
    myvars['callbackStep'] = 2
    node.setData('storeDataAsFile',0,True)
    fabricClient.flush()
    node.setData('url',0,'testfile://TMP/SubDir/testfile4.txt')
    fabricClient.flush()
    node.evaluate()
  elif myvars['callbackStep'] == 2:
    resourceResult = node.getData('resource',0)
    resourceFileContent = fabricClient.IO.getTextFileContent(resourceResult['dataExternalLocation'])
    print('Resource file content: ' + resourceFileContent)

    fabricClient.close()
  else:
    print('success callback: Unexpected')
    fabricClient.close()

node.addOnLoadSuccessCallback( loadSuccessCallback )

node.setData('url',0,'badURL')
node.evaluate()

